sorting/count_inversions.cpp

```diff
@@ -0,0 +1,198 @@
+/**
+ * @file
+ * @brief Counting Inversions using Merge Sort
+ *
+ * Program to count the number of inversions in an array
+ * using merge-sort.
+ * 
+ * The count of inversions help to determine how close the array
+ * is to being sorted in ASCENDING order.
+ * 
+ * two elements a[i] and a[j] form an inversion if a[i] > a[j] and i < j
+ * 
+ * Time Complexity --> O(n.log n)
+ * Space Complexity --> O(n) ; additional arrat temp[1..n]
+ * Algorithm:
+ *   1. The idea is similar to merge sort, divide the array into two equal or almost 
+ *      equal halves in each step until the base case is reached.
+ *   2. Create a function merge that counts the number of inversions when two halves of
+ *      the array are merged, create two indices i and j, i is the index for first half 
+ *      and j is an index of the second half. if a[i] is greater than a[j], then there are (mid – i)  
+ *      inversions, Because left and right subarrays are sorted, so all the remaining elements 
+ *      in left-subarray (a[i+1], a[i+2] … a[mid]) will be greater than a[j].
+ *   3. Create a recursive function to divide the array into halves and find the answer by summing 
+ *      the number of inversions is the first half, number of inversion in the second half and 
+ *      the number of inversions by merging the two.
+ *   4. The base case of recursion is when there is only one element in the given half.
+ *   5. Print the answer
+ * 
+ */
+#include <iostream>
+#include <vector>
+
+namespace sorting   {
+namespace inversion {
+
+int mergeSort(int* arr, int* temp, int left, int right);
+int merge(int* arr, int* temp, int left, int mid, int right);
+int countInversion(int* arr, const int size);
+void show(int* arr, const int size);
+
+/**
+ * Funtion countInversion() returns the number of inversion
+ * present in the input array. Inversions are an estimate of
+ * how close or far off the array is to being sorted.
+ * 
+ * Number of inversions in a sorted array is 0.
+ * Number of inversion in an array[1...n] sorted in 
+ * non-ascending order is n(n-1)/2, since each pair of elements
+ * contitute an inversion.
+ * 
+ * @param arr   - array, data member of std::vector<int>, input for counting inversions
+ * @param array_size    - number of elementa in the array
+ * 
+ */
+int countInversion(int* arr, const int size){
+    std::vector<int> temp;
+    temp.reserve(size);
+    temp.assign(size,0);
+    return mergeSort(arr, temp.data(), 0, size-1);
+}
+
+/**
+ * 
+ * The mergeSort() function implements Merge Sort, a
+ * Divide and conquer algorithm, it divides the input
+ * array into two halves and calls itself for each 
+ * sub-array and then calls the merge() function to 
+ * merge the two halves.
+ * 
+ * @param arr   - array to be sorted
+ * @param temp  - merged resultant array
+ * @param left  - lower bound of array
+ * @param right - upper bound of array
```

Comment:
Missing return statement documentation.
```suggestion
 * @param right - upper bound of array
 * @returns description
```